### PR TITLE
PIPE-3873 Code architectural cleanup

### PR DESCRIPTION
**Pull request by elliotforbes**
Generally improves the architectural design of the config commands to make them more easily testable.

I've extracted out the config commands into a separate package which is purely responsible for config validation and processing. It delegates all handling of various flag parsing to the `cmd` package which then passes them in as options.

This provides a better separation of concerns between the two packages and also means it's far easier to unit test the code and set up test cases as can be seen in the expanded unit test coverage.

There are still a few more optimisations that can be done, such as consolidating the `loadYaml` function into a separate package and refactoring the orb commands to use a single definition for this, but I've opted to focus purely on the config domain for now and improving that.

This takes coverage of the config package up from 0% to a far healthier:

```
github.com/CircleCI-Public/********-cli/config	0.258s	coverage: 63.1% of statements
```

# Checklist

=========

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked for similar issues and haven't found anything relevant.
- [x] This is not a security issue (which should be reported here: https://circleci.com/security/)
- [x] I have read [Contribution Guidelines](https://github.com/CircleCI-Public/circleci-cli/blob/main/CONTRIBUTING.md).

### Internal Checklist
- [x] I am requesting a review from my own team as well as the owning team
- [x] I have a plan in place for the monitoring of the changes that I am making (this can include new monitors, logs to be aware of, etc...)

## Changes

=======

- Migrates the config process and validate commands into a cleaner structure that enables us to far more easily read and test the components going forward.

## Validation

=======

- [x] Config compiles successfully
- [x] Config returns appropriate error
- [x] Config compiles successfully with k9s host
- [x] Config compiles successfully with private orb resolution
- [x] Config compile fails when no token for private orb resolution
- [x] Config compile fails when no org for private orb resolution
- [x] Config compiles successfully with private orb and private host resolution
- [x] Config compile fails with private orb and private host

### CODE REVIEW

File: api/rest/client.go [hunk 1]
```diff
@@ -17,14 +17,14 @@ import (
...
 
 type Client struct {
-	baseURL     *url.URL
+	BaseURL     *url.URL
 	circleToken string
 	client      *http.Client
 }
 
-func New(baseURL *url.URL, token string, httpClient *http.Client) *Client {
+func New(BaseURL *url.URL, token string, httpClient *http.Client) *Client {
```
[the remark this answers is not in the document]
Comment (elliotforbes) [reply api/rest/client.go:25]: I've made this public in order to allow for some explicit tests in `config_test.go`. Open to suggestions on a better approach that enables us to check the config constructor works as intended

File: config/config.go [hunk 2]
```diff
@@ -33,77 +67,73 @@ type Options struct {
...
 
 	if len(configCompilationResp.Errors) > 0 {
-		return nil, errors.New(fmt.Sprintf("config compilation contains errors: %s", configCompilationResp.Errors))
+		return nil, fmt.Errorf("config compilation contains errors: %s", configCompilationResp.Errors)
 	}
 
 	return configCompilationResp, nil
 }
+
+// #nosec
```
[the remark this answers is not in the document]
Comment (elliotforbes) [reply config/config.go]: Good spot, I'll tidy this up!

File: cmd/config.go [hunk 3]
```diff
@@ -98,12 +79,22 @@ func newConfigCommand(config *settings.Config) *cobra.Command {
...
-			return processConfig(opts, cmd.Flags())
+		RunE: func(cmd *cobra.Command, args []string) error {
+			compiler := config.New(globalConfig)
+			pipelineParamsFilePath, _ := cmd.Flags().GetString("pipeline-parameters")
+			orgID, _ := cmd.Flags().GetString("org-id")
+			orgSlug, _ := cmd.Flags().GetString("org-slug")
+			configPath := config.DefaultConfigPath
+			if len(args) >= 1 {
+				configPath = args[0]
+			}
```
[the remark this answers is not in the document]
Comment (elliotforbes) [reply cmd/config.go]: I think there should be some consistency here, I've updated with your suggestion

File: api/rest/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (elliotforbes) [reply api/rest/client.go]: Ahh I see, I've just corrected this now!

**Review by abdelDriowya — APPROVED** [review]
LGTM

File: cmd/config.go [hunk 4]
```diff
@@ -75,12 +43,24 @@ func newConfigCommand(config *settings.Config) *cobra.Command {
...
-			return validateConfig(opts, cmd.Flags())
+		RunE: func(cmd *cobra.Command, args []string) error {
+			compiler := config.New(globalConfig)
+			orgID, _ := cmd.Flags().GetString("org-id")
+			orgSlug, _ := cmd.Flags().GetString("org-slug")
+			path := config.DefaultConfigPath
+			if configPath != "" {
+				path = configPath
+			}
+			if len(args) == 1 {
```
Comment (corinnesollows) [inline cmd/config.go:54]: Are we only ever expecting one arg?

Comment (elliotforbes) [reply cmd/config.go:54]: As it stands yes based on the help output for the validate and process commands:

```
circleci config validate <path> [flags]
```

Validating this locally outputs: `Error: accepts at most 1 arg(s), received 2`

File: cmd/config.go [hunk 5]
```diff
@@ -138,115 +129,8 @@ func newConfigCommand(config *settings.Config) *cobra.Command {
...
-	}
-
-	fmt.Print(response.OutputYaml)
-	return nil
-}
-
-func packConfig(opts configOptions) error {
-	tree, err := filetree.NewTree(opts.args[0])
+func packConfig(args []string) error {
+	tree, err := filetree.NewTree(args[0])
```
Comment (corinnesollows) [inline cmd/config.go:133]: Should we verify that len(args) > 0 ? 
Unless this is being done somewhere I did not find

Comment (elliotforbes) [reply cmd/config.go:133]: I believe Line 37 `Args:        cobra.ExactArgs(1),` does this for us - otherwise it outputs the help output.

**Review by corinnesollows — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -176,3 +176,8 @@ Development instructions for the CircleCI CLI can be found in [HACKING.md](HACKI
 ## More
 
 Please see the [documentation](https://circleci-public.github.io/circleci-cli) or `circleci help` for more.
+
+
+## Version Compatibility 
+
+As of homebrew version `0.1.24705` - we no longer support Server 3.x instances. In order to upgrade the CLI to the latest version, you'll need to update your instance of server to 4.x.
```
Comment (ryanwohara) [inline README.md:183]: Users don't have to use homebrew to install the CLI. Including the package manager could be confusing to end users.

```suggestion
As of version `0.1.24705` - we no longer support Server 3.x instances. In order to upgrade the CLI to the latest version, you'll need to update your instance of server to 4.x.
```

Is `0.1.24705` the last supported version of server 3.x instances? Or do we need to update this to include a usable version?

Comment (elliotforbes) [reply README.md]: `0.1.23845` will be the last support version for server 3.x. 

I'll update this to include this information.

**Review by ryanwohara — APPROVED** [review]

**Comment by wyardley** [comment]
@elliotforbes are you strongly tied to the format here (with tabs and no newlines) vs. the one I introduced in #861 to resolve #860?

The changes here make the output pretty messy / hard to parse... current:
```
Validating config with following values
	git.base_revision:	2496ddbc7467f75ac6fd38d88396381c306f01de	id:	00000000-0000-0000-0000-000000000001	number:	1	project.git_url:	https://github.com/wyardley/circleci-cli	project.type:	github	git.tag:		git.branch:	develop	git.revision:	2496ddbc7467f75ac6fd38d88396381c306f01deNo org id or slug has been provided
```

Would you accept a PR to switch back to the other format?

(I also see "No org id or slug has been provided" even when no private orbs are used, and the way of specifying slugs seems to have changed; I can open separate issues for those if someone can take a look)(

**Comment by elliotforbes** [comment]
@wyardley  - no, I'm not strongly tied, let's get the improvements merged in and we can cut a new release - I thought I'd carried forward your changes but there were a lot of other moving parts in that PR.
